### DISCUSSION TITLE
A general/code documentation for the code

### DISCUSSION THREAD

**Discussion by CybernetHacker14** [Keriew/augustus#631] [General]
- I wanted to experiment how the dynamics change for my city, if I enable cart-pushers to drop and store goods in markets as well, so that the market ladies don't go out searching for resources. I was taking a look at the corresponding code files, and trying to make a sense of the functionality and what goes through what.
- I think a rough code documentation of sorts will help a lot of modders/gamers to tweak the game in their own vision. Would that be possible?
- If not, can I get some help in implementing something like I have mentioned it above?

**Comment by crudelios** [comment]
Hi, 

You're right, there is no documentation for the code. Coders rarely like to document their code and we're no exception ehehe.

Now, I'm your case, you need to look at `src/figuretype/cartpusher.c`. The figures work a bit like a state machine, they have a specific action that they follow and when they finish it they change their action or "die" (disappear) if they have no further meaningful actions left.

In order to add the markets to the cartpusher logic, you should look at `determine_cartpusher_destination`. Just beware that the logic is probably slightly different than the currently existing buildings.

Let me know if you need more help!

**Reply by CybernetHacker14** [reply]
@crudelios Thanks boss. I wasn't expecting any reply on this, but thanks for helping me out on this one. It will sure be a good exercise for C and modding for me.

**Reply by CybernetHacker14** [reply]
The `action.h` header file has an enum, with values like `FIGURE_ACTION_22_CARTPUSHER_DELIVERING_TO_XXXXX`. If a new action, or a destination needs to be added, does another entry needs to be added to this enum list? Or can any existing enum be used?

The reason for this question is, I am not sure whether those enum values correspond to any walker sprites or animations, etc.

**Reply by crudelios** [reply]
Technically you can use an existing enum that isn't used by that figure type. However, in terms of coding organization, it's best if you use a new one.

That being said, if you are just changing things for yourself, do what you feel it's best for you 😉

**Comment by CybernetHacker14** [comment]
Okay, some progress on this area. I am able to set the destination for the cart-pushers, and it is showing up in the dialog box that opens when we right click on the cart-pusher. However, they are stuck in the same place, and not moving at all. 

Some more debugging and breakpoints returned the fact that there's not `routing_path_id` assigned to the figure. This is because the `canTravel` value in `routce.c` is setting as 0. 

Further deep-diving in the `map_routing_citizen_can_travel_over_road_garden` function in `routing.c` shows that the `dst_offset` variable gets some absurd value as 10103. 

Checking the return value of this statement ` terrain_land_citizen.items[dst_offset]` outputs the value as -1, which corresponds to the enum `CITIZEN_N1_BLOCKED`.

Can I get some more insights on this, and as to where or what am I doing wrong?

**Reply by crudelios** [reply]
The `dst_offset` doesn't sem off, it can go from `0` to about `26000`. You probably need to set the destination to the road closest to the market, and not the market itself. Check the market ladies' code for returning to the market for proper offsets, if you use that one it should start working.

**Reply by CybernetHacker14** [reply]
Alright, lemme see it.

**Reply by CybernetHacker14** [reply]
![Screenshot_1](https://user-images.githubusercontent.com/26383905/179363722-46d1b682-de5a-4aae-abfb-3f5eb6164f41.png)
![Screenshot_2](https://user-images.githubusercontent.com/26383905/179363727-e508c0d9-511f-4fc2-ba1d-d0c90dbc09a1.png)

Finally, I am seeing some change, and the goods stocks in the market have risen tremendously. The above two pictures showcase that in a place where 4 large insulae never evolved beyond their limits, after these changes, got turned into a luxury palace almost immediately.

Thanks for your help.

One last question, how do I make a release build of this and test it out without Visual Studio?

**Reply by crudelios** [reply]
To make a release build, at the top of Visual Studio, where it says `x64-Debug`, click there and then select `Manage Configurations...`. On the window that opens, on the left side, you should see an option `x64-Debug` and a `+` button close to it. Click the button and add the `x64-Release` configuration, then select it from the drop down list. Compile everything then the executable should be at `<project folder>\out\build\x64-Release`.
